File: prover-android/app/src/main/java/org/nachweis/prover/net/RelayClient.kt
package org.nachweis.prover.net

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/** Blind relay of the verifier service (docs/blind-relay.md): request, status, one-time pickup. */
class RelayClient(private val http: OkHttpClient = defaultClient()) {
    data class Session(
        val sessionId: String,
        val requestUri: String,
        val openid4vpUri: String,
        val nonce: String,
        val boundAddress: String,
        val pickupUrl: String,
        val pickupToken: String,
        val statusUrl: String,
    )

    data class Pickup(val jwe: String, val nonce: String, val boundAddress: String, val receivedAt: Long)

    class HttpException(val code: Int, body: String) : IOException("HTTP $code: ${body.take(300)}")

    fun createRequest(baseUrl: String, clientJwk: JSONObject, boundAddress: String, challengeHex: String, redirectUri: String? = null): Session {
        val body = JSONObject()
            .put("client_jwk", clientJwk)
            .put("bound_address", boundAddress)
            .put("challenge", if (challengeHex.startsWith("0x")) challengeHex else "0x$challengeHex")
        if (!redirectUri.isNullOrBlank()) body.put("redirect_uri", redirectUri)
        val json = postJson("${baseUrl.trimEnd('/')}/relay/request", body)
        return Session(
            sessionId = json.getString("session_id"),
            requestUri = json.getString("request_uri"),
            openid4vpUri = json.getString("openid4vp_uri"),
            nonce = json.getString("nonce"),
            boundAddress = json.optString("bound_address", boundAddress),
            pickupUrl = json.getString("pickup_url"),
            pickupToken = json.getString("pickup_token"),
            statusUrl = json.getString("status_url"),
        )
    }

    /** pending | responded | picked_up */
    fun status(statusUrl: String): String = getJson(statusUrl, emptyMap()).getString("status")

    fun pickup(pickupUrl: String, pickupToken: String): Pickup {
        val json = getJson(pickupUrl, mapOf("X-Pickup-Token" to pickupToken))
        return Pickup(
            jwe = json.getString("jwe"),
            nonce = json.optString("nonce"),
            boundAddress = json.optString("bound_address"),
            receivedAt = json.optLong("received_at"),
        )
    }

    private fun postJson(url: String, body: JSONObject): JSONObject {
        val req = Request.Builder().url(url).post(body.toString().toRequestBody(JSON)).build()
        return execute(req)
    }

    private fun getJson(url: String, headers: Map<String, String>): JSONObject {
        val b = Request.Builder().url(url).get()
        headers.forEach { (k, v) -> b.header(k, v) }
        return execute(b.build())
    }

    private fun execute(req: Request): JSONObject {
        http.newCall(req).execute().use { resp ->
            val text = resp.body?.string() ?: ""
            if (!resp.isSuccessful) throw HttpException(resp.code, text)
            return JSONObject(text)
        }
    }

    companion object {
        val JSON = "application/json; charset=utf-8".toMediaType()
        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .build()
    }
}

File: prover-ios/NachweisProver/Relay/RelayClient.swift
import Foundation

/// The verifier's blind relay (nachweis-verifier-relay docs/blind-relay.md).
struct RelayClient {
    let baseURL: URL

    struct CreatedRequest: Decodable {
        let session_id: String
        let request_uri: String
        let openid4vp_uri: String
        let nonce: String
        let bound_address: String
        let pickup_url: String
        let pickup_token: String
        let status_url: String
        let response_code: String?
    }

    struct Status: Decodable { let status: String }

    struct Pickup: Decodable {
        let jwe: String
        let nonce: String
        let bound_address: String
        let received_at: Int64
    }

    /// POST /relay/request. `challengeHex` is the 32-byte challenge the bridge
    /// session was created with (nonce = sha256(address20 || challenge)).
    func createRequest(clientJWK: [String: String], boundAddress: String, challengeHex: String, redirectURI: String?) async throws -> CreatedRequest {
        var body: [String: Any] = [
            "client_jwk": clientJWK,
            "bound_address": boundAddress,
            "challenge": challengeHex.hasPrefix("0x") ? challengeHex : "0x" + challengeHex,
        ]
        if let r = redirectURI, !r.isEmpty { body["redirect_uri"] = r }
        return try await HTTP.postJSON(baseURL.appendingPathComponent("relay/request"), body: body)
    }

    func status(sessionID: String) async throws -> String {
        let s: Status = try await HTTP.getJSON(baseURL.appendingPathComponent("relay/status/\(sessionID)"))
        return s.status
    }

    /// One-time pickup with the token; the relay drops the JWE afterwards.
    func pickup(sessionID: String, token: String) async throws -> Pickup {
        try await HTTP.getJSON(baseURL.appendingPathComponent("relay/response/\(sessionID)"), headers: ["X-Pickup-Token": token])
    }
}

enum HTTP {
    struct Error: Swift.Error, LocalizedError {
        let status: Int
        let body: String
        var errorDescription: String? { "HTTP \(status): \(body.prefix(300))" }
    }

    static func postJSON<T: Decodable>(_ url: URL, body: [String: Any], headers: [String: String] = [:]) async throws -> T {
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(req)
    }

    static func getJSON<T: Decodable>(_ url: URL, headers: [String: String] = [:]) async throws -> T {
        var req = URLRequest(url: url)
        headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
        return try await send(req)
    }

    private static func send<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp) = try await URLSession.shared.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else { throw Error(status: status, body: String(decoding: data, as: UTF8.self)) }
        return try JSONDecoder().decode(T.self, from: data)
    }
}
